### PR TITLE
Enhanced relid generation

### PR DESCRIPTION
**Pull request by kecso**
- now the relid is generated as short as possible (character pool is numbers, small, and large cap letters) starting from a single character
- enhanced relid collision have been implemented (new use-cases during merge)
- library update have also been corrected (if an item have been removed only from a META sheet, but not the whole project)
- necessary tests have been added

### CODE REVIEW

File: src/common/util/random.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
+/*globals define*/
+/*jshint node: true, browser: true, bitwise: false*/
+
+/**
+ * @author kecso / https://github.com/kecso
+ *
+ * collection of functions that uses random Numbers in WebGME
+ */
+
+define(['./chance.min'], function (ChanceJs) {
```
Comment (pmeijer) [inline src/common/util/random.js]: This file should have a random.spec.js.
Try things like generateRelid when object contains different keys.

File: src/common/util/random.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    'use strict';
+
+    function random() {
```
Comment (pmeijer) [inline src/common/util/random.js]: Random (it's used with new).

File: src/common/util/random.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        var guidPool = '0123456789abcdef',
+            relidPool = '0123456789qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM',
+            excludeList = ['atr', 'reg', 'ovr'],
+            maxTry = 2,
+            chance = new ChanceJs(),
```
Comment (pmeijer) [inline src/common/util/random.js]: Are you sure chance should be initialized? It should suffice to remove this and just require it as chance.

File: src/common/util/random.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            randFunction = function (max) {
+                //return Math.floor(Math.random() * max);
+                return chance.natural({max:max-1});
+            },
+            validRelid = generateRelidRegexp();
+
+        function generateRelidRegexp() {
```
Comment (pmeijer) [inline src/common/util/random.js]: Maybe add an _ if it's not exposed alt add the other directly on this.

File: src/common/core/corediff.js [hunk 5]
```diff
@@ -10,10 +10,13 @@
...
     'common/core/tasync',
     'common/util/assert',
-    'common/regexp'
-], function (CANON, TASYNC, ASSERT, REGEXP) {
+    'common/regexp',
+    'common/util/random'
+], function (CANON, TASYNC, ASSERT, REGEXP, Radom) {
     'use strict';
 
+    var generateRelid = (new Radom()).generateRelid;
```
Comment (pmeijer) [inline src/common/core/corediff.js]: Random

File: src/common/util/random.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                if (tries >= maxTry) {
+                    tries = 0;
+                    length += 1;
+                }
+                relid = '';
+                for (i = 0; i < length; i += 1) {
+                    relid += getRandomCharacter(relidPool);
+                }
+                tries += 1;
+            } while (object[relid] !== undefined);
```
Comment (pmeijer) [inline src/common/util/random.js]: Would hasOwnProperty work? Or are there cases where the object gets assigned an undefined?

File: src/common/core/guidcore.js [hunk 7]
```diff
@@ -46,7 +49,8 @@ define([
             //TODO we always should know what structure we should expect as a relid -
             // now we think it is a number so it can be converted to 0xsomething
             var relid = _core.getRelid(node);
-            relid = Number(relid);
+            //relid = Number(relid);
+            relid = relidToInteger(relid);
```
Comment (pmeijer) [inline src/common/core/guidcore.js]: Why does it need to be converted?

File: src/common/core/corediff.js [hunk 8]
```diff
@@ -811,6 +813,89 @@ define(['common/util/canon',
...
+
+                    //now the actual place switching
+                    parent = getPathOfDiff(globalDiff, getParentPath(path));
+                    parent[newRelid] = nodeDiff;
+                    delete parent[relid];
+                    dst2src[newPath] = dst2src[path];
+                    delete dst2src[path];
+                    src2dst[dst2src[newPath]] = newPath;
+                } else {
+                    logger.warning('there is a guid mismatch among the two diffs: ' +
```
Comment (pmeijer) [inline src/common/core/corediff.js]: Is this the unexpected, what are the consequences? Maybe throw an error in an else at 863?

File: src/common/core/corediff.js [hunk 9]
```diff
@@ -811,6 +813,89 @@ define(['common/util/canon',
...
+                        diffBase.guid + ' vs ' + diffExtension.guid);
+                }
+            }
+
+            //recursive calls - only if there were no replacement due to collision
+            if (!needToRelocate) {
```
Comment (pmeijer) [inline src/common/core/corediff.js]: I would only use ! when there's no bool (you seem to do this elsewhere too)

File: src/common/core/corediff.js [hunk 10]
```diff
@@ -843,11 +928,41 @@ define(['common/util/canon',
         }
 
         function makeInitialContainmentChanges(node, diff) {
-            var relids = getDiffChildrenRelids(diff),
+            var getOrderedRelids = function () {
```
Comment (pmeijer) [inline src/common/core/corediff.js]: Does this need to be within this scope?

File: src/common/core/corerel.js [hunk 11]
```diff
@@ -5,16 +5,21 @@
...
     // ----------------- RELID -----------------
 
-    var ATTRIBUTES = 'atr';
-    var REGISTRY = 'reg';
-    var OVERLAYS = 'ovr';
-    var COLLSUFFIX = '-inv';
+    var ATTRIBUTES = 'atr',
+        REGISTRY = 'reg',
+        OVERLAYS = 'ovr',
+        COLLSUFFIX = '-inv';
```
Comment (pmeijer) [inline src/common/core/corerel.js]: Constants of core...

File: src/common/core/coretree.js [hunk 12]
```diff
@@ -8,45 +8,15 @@
...
-    //        do {
-    //            relid = (nextRelid += -1);
-    //        } while (data[relid] !== undefined);
-    //
-    //        return '' + relid;
-    //    };
-    //}
-
-    var rootCounter = 0;
+    var HASH_REGEXP = new RegExp('#[0-9a-f]{40}'),
```
Comment (pmeijer) [inline src/common/core/coretree.js]: This should be in common regexp...
(You can keep the old one for now there too, call this SHA1 or something like that)

Comment (pmeijer) [inline src/common/core/coretree.js]: Wait is it even used? Are both used? REGEXP.DB_HASH

File: src/common/core/guidcore.js [hunk 13]
```diff
@@ -85,103 +85,108 @@ define([
...
-            _core.setAttribute(newNode, OWN_GUID, xorGuids(_core.getMiddleGuid(parent),
-                xorGuids(oldGuid, getRelidGuid(newNode))));
-
-            return newNode;
-        };
+            var oldGuid = _core.getGuid(node);
+
+            node = _innerCore.moveNode(node, parent);
+
+            _core.setAttribute(node, OWN_GUID,
```
Comment (pmeijer) [inline src/common/core/guidcore.js]: Isn't this the same as `setDataGuid`, line 88

File: src/common/core/guidcore.js [hunk 14]
```diff
@@ -6,14 +6,17 @@
```
Comment (pmeijer) [inline src/common/core/guidcore.js]: Why does the GUID core have to change?

File: src/common/core/users/serialization.js [hunk 15]
```diff
@@ -1177,10 +1174,16 @@ define(['common/util/assert', 'blob/BlobConfig'], function (ASSERT, BlobConfig)
...
                     var oldMemberGuids = Object.keys(oldSheets[name]),
-                        i;
+                        newMemberGuids = Object.keys(newSheets[name]);
+
+                    for (i = 0; i < newMemberGuids.length; i += 1) {
+                        if (oldMemberGuids.indexOf(newMemberGuids[i]) === -1 && nodes[newMemberGuids[i]]) {
+                            //it was only removed from the sheet
+                            core.delMember(root, name, newMemberGuids[i]);
+                        }
+                    }
```
Comment (pmeijer) [inline src/common/core/users/serialization.js]: Are these changes related?

File: src/common/regexp.js [hunk 16]
```diff
@@ -18,6 +19,7 @@ define([], function () {
 
     return {
         HASH: HASH,
+        DB_HASH: DB_HASH,
```
Comment (pmeijer) [inline src/common/regexp.js]: Oh, here he is :)

File: src/common/core/users/serialization.js (hunk 15 above)
Comment (kecso) [inline src/common/core/users/serialization.js]: not entirely.
during guid checks it turned out, that we have a missing use-case in our portfolio

File: src/common/core/corerel.js [hunk 17]
```diff
@@ -47,6 +32,16 @@ define(['common/util/assert', 'common/core/coretree', 'common/core/tasync'], fun
...
         logger.debug('initialized');
 
+        function isPointerName(name) {
+            ASSERT(typeof name === 'string');
+            //TODO this is needed as now we work with modified data as well
+            if (name === '_mutable') {
+                return false;
+            }
+            return name.slice(-coretree.constants.COLLECTION_NAME_SUFFIX.length) !==
+                coretree.constants.COLLECTION_NAME_SUFFIX;
```
Comment (pmeijer) [inline src/common/core/corerel.js]: This checks that it doesn't end with COLLECTION_NAME_SUFFIX, right?
